src/passes/Directize.cpp

```diff
@@ -89,14 +130,48 @@ struct FunctionDirectizer : public WalkerPass<PostWalker<FunctionDirectizer>> {
 
   bool changedTypes = false;
 
-  void replaceWithUnreachable(CallIndirect* call) {
-    Builder builder(*getModule());
-    for (auto*& operand : call->operands) {
-      operand = builder.makeDrop(operand);
+  // Create a direct call for a given list of operands, an expression which is
+  // known to contain a constant indicating the table offset, and the relevant
+  // table. If we can see that the call will trap, instead return an
+  // unreachable.
+  Expression* makeDirectCall(const std::vector<Expression*>& operands,
+                             Expression* c,
+                             const TableUtils::FlatTable& flatTable,
+                             CallIndirect* original) {
+    Index index = c->cast<Const>()->value.geti32();
+
+    // If the index is invalid, or the type is wrong, we can
+    // emit an unreachable here, since in Binaryen it is ok to
+    // reorder/replace traps when optimizing (but never to
+    // remove them, at least not by default).
+    if (index >= flatTable.names.size()) {
+      return replaceWithUnreachable(operands);
+    }
+    auto name = flatTable.names[index];
+    if (!name.is()) {
+      return replaceWithUnreachable(operands);
+    }
+    auto* func = getModule()->getFunction(name);
+    if (original->sig != func->getSig()) {
```

Comment:
Might as well use a subtype check here to make this more future-proof while we're changing it. wdyt?

Reply:
Good idea to do that, but I'd prefer to do it in a followup.

Reply:
Actually, can we do this check here? `HeapType::isSubType` would check nominally, but this must be a structural check.

Reply:
Or did you mean to do a subtyping check on each item of the params and results separately, or something like that?

Reply:
Oh yeah we would need a structural signature subtype check even in nominal mode. I guess we don't have an API for that yet since we don't actually support function subtyping.